incubator/declarative-linking/src/main/java/org/glassfish/jersey/linking/InjectLinkMethodDescriptor.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+
+package org.glassfish.jersey.linking;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.ws.rs.core.Link;
+
+import org.glassfish.jersey.linking.mapping.ResourceMappingContext;
+
+/**
+ * Utility class for working with {@link InjectLink} annotated methods.
+ *
+ * @author Ryan Peterson
+ */
+class InjectLinkMethodDescriptor extends MethodDescriptor implements InjectLinkDescriptor {
+
+    private InjectLink link;
+    private Class<?> type;
+    private Map<String, String> bindings;
+
+    /**
+     * TODO javadoc.
+     */
+    public InjectLinkMethodDescriptor(Method m, InjectLink l, Class<?> t) {
+        super(t, m);
+        link = l;
+        type = t;
+        bindings = new HashMap<>();
+        for (Binding binding : l.bindings()) {
+            bindings.put(binding.name(), binding.value());
+        }
+    }
+
+    /**
+     * TODO javadoc.
+     */
+    public void setPropertyValue(Object instance, URI uri) {
+        setAccessibleMethod(getter);
+        try {
+
+            Object value;
+            if (URI.class.equals(type)) {
+                value = uri;
+            } else if (Link.class.isAssignableFrom(type)) {
+
+                // Make a link with the correct bindings
+                value = getLink(uri);
+            } else if (String.class.equals(type)) {
+                value = uri.toString();
+            } else {
+                throw new IllegalArgumentException("Field type " + type + " not one of supported String,URI and Link");
+            }
+
+            setter.invoke(instance, value);
+        } catch (IllegalArgumentException | IllegalAccessException | InvocationTargetException ex) {
+            Logger.getLogger(InjectLinkFieldDescriptor.class.getName()).log(Level.SEVERE, null, ex);
+        }
+    }
+
+    /**
+     * TODO javadoc.
+     */
+    public InjectLink.Style getLinkStyle() {
+        return link.style();
+    }
+
+    /**
+     * TODO javadoc.
+     */
+    public String getLinkTemplate(ResourceMappingContext rmc) {
+        return InjectLinkFieldDescriptor.getLinkTemplate(rmc, link);
```

Comment:
I know, that this call is technically OK - it will work, but it does not really feel right. 

Actually the entire class again shares a huge amount of code with InjectLinkFieldDescriptor. I think it would really make sense to extract the common parts into abstract superclass or do some other refactoring that reduces the amount of repeating code.